### ISSUE TITLE
Model.CSVParser.readRowCount(CSVParser.java:112)

### ISSUE THREAD

**Issue by sathyajackson** [codingchili/excelastic#71]
Hello,

Once again when I try to upload the monthly bill using the tool. I ended up with the below error.

Nov 22, 2018 10:43:16 AM io.vertx.core.impl.BlockedThreadChecker
WARNING: Thread Thread[vert.x-worker-thread-2,5,main] has been blocked for 284295 ms, time limit is 60000
io.vertx.core.VertxException: Thread blocked
        at app//com.codingchili.Model.CSVParser.readRowCount(CSVParser.java:112)
        at app//com.codingchili.Model.CSVParser.initialize(CSVParser.java:96)
        at app//com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
        at app//com.codingchili.Controller.Website$$Lambda$229/226717646.handle(Unknown Source)
        at app//io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
        at app//io.vertx.core.impl.ContextImpl$$Lambda$195/110018830.run(Unknown Source)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base@10.0.2/java.lang.Thread.run(Thread.java:844)

**Comment by codingchili** [comment]
Hi,

If it's a really large file you will see these warnings.
It's not an issue unless there are any errors in the log.
This is just to indicate that a blocking task has been blocking for a very long time.

Does it stay blocked like this forever? Or does it eventually start importing?

**Comment by sathyajackson** [comment]
I get this error message only after 100 % is done, but the problem is, i dont find the index or the data in kibana.... :(

**Comment by codingchili** [comment]
Aha, can you send me the full log from excelastic and elasticsearch?

**Comment by sathyajackson** [comment]
The one which I pasted is why I get from excelastic... I don't see any thing more than that in elastic search.

I will rerun and let you know

**Comment by codingchili** [comment]
Excelastic should be logging more than that exception. If it's not then that's another issue.

**Comment by sathyajackson** [comment]
I tired again and ended up with below error on excelsitic console 

java.nio.BufferOverflowException
	at java.base/java.nio.Buffer.nextPutIndex(Buffer.java:649)
	at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:209)
	at com.codingchili.Model.CSVParser.readHeaders(CSVParser.java:133)
	at com.codingchili.Model.CSVParser.initialize(CSVParser.java:97)
	at com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:844)

on the elasticsearch logs 

elasticsearch log 

2018-11-28T11:30:10,235][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:30:40,262][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:31:10,376][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:31:40,377][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:32:10,382][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:32:40,383][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:33:10,384][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:33:40,388][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:34:10,392][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:34:40,393][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node
[2018-11-28T11:35:10,395][INFO ][o.e.c.r.a.DiskThresholdMonitor] [mOS9xvx] low disk watermark [85%] exceeded on [mOS9xvxQSzyCxnCZAjQ1JA][mOS9xvx][/var/lib/elasticsearch/nodes/0] free: 18.9gb[14.6%], replicas will not be assigned to this node

I have more than 20 G of free space on the disk so, I think this is not what we are looking for.

on gc.log 

[2018-11-28T06:07:20.735+0000][12930][safepoint    ] Application time: 2.9923693 seconds
[2018-11-28T06:07:20.735+0000][12930][gc,start     ] GC(44) Pause Young (Allocation Failure)
[2018-11-28T06:07:20.735+0000][12930][gc,task      ] GC(44) Using 4 workers of 4 for evacuation
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) Desired survivor size 17432576 bytes, new threshold 6 (max threshold 6)
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) Age table with threshold 6 (max threshold 6)
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   1:    7926048 bytes,    7926048 total
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   2:      21392 bytes,    7947440 total
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   3:       4216 bytes,    7951656 total
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   4:     291088 bytes,    8242744 total
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   5:      44176 bytes,    8286920 total
[2018-11-28T06:07:20.742+0000][12930][gc,age       ] GC(44) - age   6:       3096 bytes,    8290016 total
[2018-11-28T06:07:20.742+0000][12930][gc,heap      ] GC(44) ParNew: 277641K->10942K(306688K)
[2018-11-28T06:07:20.742+0000][12930][gc,heap      ] GC(44) CMS: 205320K->205322K(707840K)
[2018-11-28T06:07:20.742+0000][12930][gc,metaspace ] GC(44) Metaspace: 85888K->85888K(1128448K)
[2018-11-28T06:07:20.742+0000][12930][gc           ] GC(44) Pause Young (Allocation Failure) 471M->211M(990M) 6.392ms
[2018-11-28T06:07:20.742+0000][12930][gc,cpu       ] GC(44) User=0.03s Sys=0.00s Real=0.00s
[2018-11-28T06:07:20.742+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0066804 seconds, Stopping threads took: 0.0000298 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Application time: 0.0028443 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0001649 seconds, Stopping threads took: 0.0000196 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Application time: 0.0003960 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0000603 seconds, Stopping threads took: 0.0000073 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Application time: 0.0000759 seconds
[2018-11-28T06:07:20.745+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0000449 seconds, Stopping threads took: 0.0000044 seconds
[2018-11-28T06:07:20.746+0000][12930][safepoint    ] Application time: 0.0000253 seconds
[2018-11-28T06:07:20.746+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0000356 seconds, Stopping threads took: 0.0000040 seconds
[2018-11-28T06:07:20.749+0000][12930][safepoint    ] Application time: 0.0031927 seconds
[2018-11-28T06:07:20.749+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0000916 seconds, Stopping threads took: 0.0000142 seconds
[2018-11-28T06:07:22.742+0000][12930][safepoint    ] Application time: 1.9934294 seconds
[2018-11-28T06:07:22.742+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0001140 seconds, Stopping threads took: 0.0000185 seconds
[2018-11-28T06:07:22.743+0000][12930][safepoint    ] Application time: 0.0001237 seconds
[2018-11-28T06:07:22.743+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0000689 seconds, Stopping threads took: 0.0000070 seconds
[2018-11-28T06:07:27.743+0000][12930][safepoint    ] Application time: 5.0002211 seconds
[2018-11-28T06:07:27.743+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0002170 seconds, Stopping threads took: 0.0000297 seconds
[2018-11-28T06:07:27.743+0000][12930][safepoint    ] Application time: 0.0001833 seconds
[2018-11-28T06:07:27.744+0000][12930][safepoint    ] Total time for which application threads were stopped: 0.0001016 seconds, Stopping threads took: 0.0000120 seconds

**Comment by sathyajackson** [comment]
Also i see these error messages on the JVM console 

Nov 28, 2018 11:34:36 AM io.vertx.core.impl.BlockedThreadChecker
WARNING: Thread Thread[vert.x-worker-thread-1,5,main] has been blocked for 572607 ms, time limit is 60000
io.vertx.core.VertxException: Thread blocked
        at app//com.codingchili.Model.CSVParser.readRowCount(CSVParser.java:112)
        at app//com.codingchili.Model.CSVParser.initialize(CSVParser.java:96)
        at app//com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
        at app//com.codingchili.Controller.Website$$Lambda$217/393893673.handle(Unknown Source)
        at app//io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
        at app//io.vertx.core.impl.ContextImpl$$Lambda$202/710828371.run(Unknown Source)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base@10.0.2/java.lang.Thread.run(Thread.java:844)

Nov 28, 2018 11:34:36 AM com.codingchili.Controller.Website lambda$onComplete$9
SEVERE: Failed to parse file '620462907849-aws-billing-detailed-line-items-with-resources-and-tags-2018-03.csv'.
java.nio.BufferOverflowException
        at java.base/java.nio.Buffer.nextPutIndex(Buffer.java:649)
        at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:209)
        at com.codingchili.Model.CSVParser.readHeaders(CSVParser.java:133)
        at com.codingchili.Model.CSVParser.initialize(CSVParser.java:97)
        at com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:844)

**Comment by sathyajackson** [comment]
Also i see the below error on the JVM console.

Nov 28, 2018 11:34:36 AM io.vertx.core.impl.BlockedThreadChecker
WARNING: Thread Thread[vert.x-worker-thread-1,5,main] has been blocked for 572607 ms, time limit is 60000
io.vertx.core.VertxException: Thread blocked
        at app//com.codingchili.Model.CSVParser.readRowCount(CSVParser.java:112)
        at app//com.codingchili.Model.CSVParser.initialize(CSVParser.java:96)
        at app//com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
        at app//com.codingchili.Controller.Website$$Lambda$217/393893673.handle(Unknown Source)
        at app//io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
        at app//io.vertx.core.impl.ContextImpl$$Lambda$202/710828371.run(Unknown Source)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base@10.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base@10.0.2/java.lang.Thread.run(Thread.java:844)

Nov 28, 2018 11:34:36 AM com.codingchili.Controller.Website lambda$onComplete$9
SEVERE: Failed to parse file 'XXXXXXX-2018-03.csv'.
java.nio.BufferOverflowException
        at java.base/java.nio.Buffer.nextPutIndex(Buffer.java:649)
        at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:209)
        at com.codingchili.Model.CSVParser.readHeaders(CSVParser.java:133)
        at com.codingchili.Model.CSVParser.initialize(CSVParser.java:97)
        at com.codingchili.Controller.Website.lambda$parse$11(Website.java:183)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:273)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:844)

**Comment by sathyajackson** [comment]
I can see the complete file has been uploaded on to file-uploads folder, but i guess its having problem when making it with elasticsearch for indexing.

If i load a small file close to 1 G or 700mb, no problem this works great but again the file which i am loading is 22G and close to 3 million rows (not that exact on this numbers, but yes its huge)

**Comment by codingchili** [comment]
Thanks a lot, this is an error that occurs during the quick parse, it's done before the import starts to ensure the whole file is parsable. This looks a lot like the error that I fixed recently, are you using 1.3.3? It's either a similar error - failing to detect end of line, or there is a line in the data that is too big for the line buffer.

**Comment by sathyajackson** [comment]
Yes I am using the latest, may be the fix which you provide doesn't support on humangous size files? 

Eager to know what's next, to make this tool really a productive one for larger files.....

**Comment by codingchili** [comment]
This is when headers are being parsed. The very first line.
Oh my, there appears to be a very embarrassing bug here.

When I added support for LF and CRLF files, I messed up when reading headers.
When reading headers we use CR as the EOL character ! This means we will never detect EOL and read the whole file as headers. That's why the buffer fills. 

I will fix this ASAP. It should work better if you convert to Windows CRLF line endings until then.

**Comment by sathyajackson** [comment]
Oh good, but I think I lost you on the last part, how to use windows CRLF.... I am kind of new to this and I have no idea about it.

If you share some docs or steps I can try it.

**Comment by codingchili** [comment]
I tried the AWS cost explorer CSV export. There are some issues with it.

- it contains headers without names, for example ",field 1, field 2" this is bad for Elasticsearch. 

Changing it so that we are generating new headers where it is missing. looks like "header_1, field 1, field 2".

- data types are being mixed for the "Total", all other lines contains a valid date yyyy-mm-dd except the one with totals, which say "Historical totals" - this cannot be parsed as a date. 

I don't know how to solve this. Maybe its possible to skip the "total" field in the cost explorer. The problem is that if ElasticSearch encounter the "string" before the "date" type - the dynamically generated mapping will be wrong.



The AWS CSV file uses LF endings, I tested it and it works with the recent changes i applied - but only if i remove the totals.

**Comment by sathyajackson** [comment]
Ah ok, any suggestions that you have to fix or alternate to load using the UI ?

Removing total from that file is not possible as it's huge and I can't open that file to edit it. 

Any technical plans on your tool to make it work or to ignore certain fields ?

**Comment by codingchili** [comment]
I need a moment to think :) it seems that Elasticsearch ignores fields that are not consistent with the mapping. But still, the problem would arise if the dynamic mapping was generated from the first line, which contains the totals.. 

Maybe we can't rely on the dynamic mapping in this case. If we set a mapping before starting the import, we can specify the field types that are valid for import. I think this will be the solution. I will try to make a new release tomorrow with the changes to line endings.

**Comment by sathyajackson** [comment]
Alrite.... Waiting for the next release to test on huge imports.

Thanks for your help...

**Comment by codingchili** [comment]
Released 1.3.4, tested with AWS CSV files. Mixed data types causes some errors in elasticsearch, uniform data types imported ok - as long as dynamic mapping has correct types.

**Comment by sathyajackson** [comment]
Will check the latest release and update you.

**Comment by sathyajackson** [comment]
Alrite I checked and I am still ended up with the same issue. I don't know but you have tried hard in helping, I think this might be a buggy file which I am trying to upload.

**Comment by codingchili** [comment]
Yep, probably. Thanks for testing.

Closing, please reopen if required.
